File: lib/image_provider.dart
import 'dart:io';

enum Resource { NETWORK, DRAWABLE, FILE, ASSET }

/// Drawable图片资源
class DrawableImageProvider extends ImageProvider<String> {
  final String drawable;

  DrawableImageProvider(this.drawable);

  @override
  Map<String, dynamic> resolve() => {"resource": drawable, "resourceType": Resource.DRAWABLE.index};
}

/// 文件图片资源
class FileImageProvider extends ImageProvider<File> {
  final File file;

  FileImageProvider(this.file);

  @override
  Map<String, dynamic> resolve() => {"resource": file.path, "resourceType": Resource.FILE.index};
}

/// 网络图片资源
class NetworkImageProvider extends ImageProvider<String> {
  final String url;

  NetworkImageProvider(this.url);

  @override
  Map<String, dynamic> resolve() => {"resource": url, "resourceType": Resource.NETWORK.index};

}

/// Asset图片资源
class AssetImageProvider extends ImageProvider<String> {
  final String asset;

  AssetImageProvider(this.asset);
  
  @override
  Map<String, dynamic> resolve() => {"resource": asset, "resourceType": Resource.ASSET.index};
  
}

abstract class ImageProvider<T> {
  Map<String, dynamic> resolve();
}

File: android/src/main/kotlin/org/axen/flutter/glide/common/ImageProvider.kt
package org.axen.flutter.glide.common

import android.content.Context
import com.bumptech.glide.RequestBuilder
import com.bumptech.glide.load.DecodeFormat
import com.bumptech.glide.load.engine.DiskCacheStrategy
import org.axen.flutter.glide.converter.AssetConverter
import org.axen.flutter.glide.converter.DrawableConverter
import org.axen.flutter.glide.converter.FileConverter
import org.axen.flutter.glide.converter.NetworkConverter
import org.axen.flutter.glide.enum.BoxFit
import org.axen.flutter.glide.enum.Resource

abstract class AbstractProvider<T>(protected val context: Context) : ImageProvider<T> {

    private val converters: MutableMap<Resource, ImageConverter<T>> = mutableMapOf()

    init { register() }

    private fun register() {
        converters[Resource.FILE] = FileConverter()
        converters[Resource.NETWORK] = NetworkConverter()
        converters[Resource.DRAWABLE] = DrawableConverter()
        converters[Resource.ASSET] = AssetConverter()
    }

    override fun provide(info: GlideImage): T {
        val converter = converters[info.resourceType]
            ?: throw RuntimeException("Not support resource type!")
        val request = converter.convert(context, fit(build(), info.fit), info.resource!!)
        val futureTarget = if (info.width > 0 && info.height > 0) {
            val w = info.width * info.scaleRatio
            val h = info.height * info.scaleRatio
            request.submit(w.toInt(), h.toInt())
        } else request.submit()

        return futureTarget.get()
    }

    abstract fun build(): RequestBuilder<T>

    companion object {
        fun <T> fit(builder: RequestBuilder<T>, fit: BoxFit): RequestBuilder<T> {
            return when(fit) {
                BoxFit.COVER -> builder.centerCrop()
                BoxFit.CONTAIN,
                BoxFit.FIT_WIDTH,
                BoxFit.FIT_HEIGHT,
                BoxFit.SCALE_DOWN -> builder.fitCenter()
                else -> builder
            }
        }
    }
}

interface ImageProvider<T> {
    fun provide(info: GlideImage) : T
}